### PR TITLE
fix: Add unsupported framework rule

### PR DESCRIPTION
**Pull request by DaveTryon**
#### Details

Issue #554 requests a way to warn users if they're using a framework that doesn't support UIA. We don't yet have a good way to do this in live mode, but a first step is to add an automated check that triggers an error based on the Windows class name. So far, the cases we've encountered are Win32 frameworks that return class names of `SunAwtFrame` or `SunAwtDialog`. Other classes could exist, so we've opted to raise the error for any Win32 object with a reported ClassName of `SunAwt*`. We use a case-sensitive regular expression here.

##### Motivation

Issue #554 

##### Context

The HowToFix text may be a little too blunt. It says "Switch your application to a framework that supports UI Automation.". I'm open to gentler ways to express this sentiment of "Sorry, you chose the wrong framework".

#### Pull request checklist

- [ ] Addresses an existing issue: #0000

### CODE REVIEW

File: src/Core/Misc/ExtensionMethods.cs [hunk 1]
```diff
@@ -647,9 +647,11 @@ public static string GetUIFramework(this IA11yElement e)
         {
             if (e == null) return null;
 
-            return (string.IsNullOrEmpty(e.Framework))
+            string framework = e.Framework;
+
+            return (string.IsNullOrEmpty(framework))
```
Comment (DaveTryon) [inline src/Core/Misc/ExtensionMethods.cs:652]: While running the unit tests, I noticed that e.Framework was being called twice. I had the choice to tell the unit tests to double their expected number of calls, or change the Extension method to only query the property once. The latter seemed like the better choice. Happy to discuss if you wish.

File: src/Core/Misc/ExtensionMethods.cs [hunk 2]
```diff
@@ -647,9 +647,11 @@ public static string GetUIFramework(this IA11yElement e)
         {
             if (e == null) return null;
 
-            return (string.IsNullOrEmpty(e.Framework))
```
Comment (RobGallo) [inline src/Core/Misc/ExtensionMethods.cs:650]: Not quite sure why this change is here.

File: src/Core/Misc/ExtensionMethods.cs (hunk 1 above)
Comment (RobGallo) [reply src/Core/Misc/ExtensionMethods.cs:652]: I think this needs a comment in the code.

File: src/Rules/Library/FrameworkDoesNotSupportUIAutomation.cs [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+using Axe.Windows.Rules.Resources;
+using System.Text.RegularExpressions;
+using static Axe.Windows.Rules.PropertyConditions.Framework;
+
+namespace Axe.Windows.Rules.Library
+{
+    [RuleInfo(ID = RuleId.FrameworkDoesNotSupportUIAutomation)]
+    class FrameworkDoesNotSupportUIAutomation : Rule
+    {
+        private readonly static string[] KnownProblematiClasses =
```
Comment (RobGallo) [inline src/Rules/Library/FrameworkDoesNotSupportUIAutomation.cs:14]: Spelling

File: src/Rules/Library/FrameworkDoesNotSupportUIAutomation.cs [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+                if (Regex.IsMatch(className, knownProblematicClass, RegexOptions.CultureInvariant))
+                    return false;
+            }
+
+            return true;
+        }
+
+        protected override Condition CreateCondition()
+        {
+            return Win32Framework;
```
Comment (RobGallo) [inline src/Rules/Library/FrameworkDoesNotSupportUIAutomation.cs:42]: Wondering if we want to narrow this down a little further so this test doesn't get run all the time. For example, are the SAWT* class types usually only on elements with a control type of Window or Pane?

File: src/Rules/Resources/Descriptions.resx [hunk 5]
```diff
@@ -426,4 +426,7 @@
   <data name="ClickablePointOffScreen" xml:space="preserve">
     <value>An element's IsOffScreen property must be true when its clickable point is off screen.</value>
   </data>
+  <data name="FrameworkDoesNotSupportUIAutomation" xml:space="preserve">
+    <value>The framework in use does not support UI Automation.</value>
```
Comment (RobGallo) [inline src/Rules/Resources/Descriptions.resx:430]:
```suggestion
    <value>The framework used to build this application does not support UI Automation.</value>
```

File: src/RulesTest/Library/FrameworkDoesNotSupportUIAutomationTests.cs [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+        [TestMethod]
+        public void Condition_FrameworkIsNotWin32_ReturnsFalse()
+        {
+            string[] nonWin32Values = { FrameworkId.DirectUI, FrameworkId.Edge, FrameworkId.InternetExplorer, FrameworkId.WinForm, FrameworkId.WPF, FrameworkId.XAML, "NotWin32" };
+
+            foreach (string nonWin32Value in nonWin32Values)
+            {
+                _elementMock.Setup(m => m.Framework)
+                    .Returns(FrameworkId.DirectUI)
```
Comment (RobGallo) [inline src/RulesTest/Library/FrameworkDoesNotSupportUIAutomationTests.cs:32]: Don't you intend to return `nonWin32Value`?

File: src/RulesTest/Library/FrameworkDoesNotSupportUIAutomationTests.cs [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+        {
+            _elementMock.Setup(m => m.Framework).Returns(FrameworkId.Win32).Verifiable();
+
+            Assert.IsTrue(Rule.Condition.Matches(_elementMock.Object));
+
+            _elementMock.Verify(m => m.Framework, Times.Once());
+        }
+
+        [TestMethod]
+        public void PassesTest_FrameworkIsWin32_ClassNameIsNotKnownProblematic_ReturnsFalse()
```
Comment (RobGallo) [inline src/RulesTest/Library/FrameworkDoesNotSupportUIAutomationTests.cs:52]: I think "IsWin32" in the name is not necessary, and could easily become out-of-date.

**Review by RobGallo — CHANGES_REQUESTED** [review]

File: src/Rules/Resources/Descriptions.resx (hunk 5 above)
Comment (DaveTryon) [reply src/Rules/Resources/Descriptions.resx]: Fixed

File: src/Rules/Library/FrameworkDoesNotSupportUIAutomation.cs (hunk 4 above)
Comment (DaveTryon) [reply src/Rules/Library/FrameworkDoesNotSupportUIAutomation.cs]: Narrowed the scope to just Windows (this is the only case that occurs in the sample files that we've received)

File: src/Rules/Library/FrameworkDoesNotSupportUIAutomation.cs (hunk 3 above)
Comment (DaveTryon) [reply src/Rules/Library/FrameworkDoesNotSupportUIAutomation.cs]: Fixed

File: src/Core/Misc/ExtensionMethods.cs (hunk 1 above)
Comment (DaveTryon) [reply src/Core/Misc/ExtensionMethods.cs:653]: Comment added

File: src/RulesTest/Library/FrameworkDoesNotSupportUIAutomationTests.cs (hunk 6 above)
Comment (DaveTryon) [reply src/RulesTest/Library/FrameworkDoesNotSupportUIAutomationTests.cs]: Yes. The first version had a hardcoded value and I forgot to update it when I added the loop

File: src/RulesTest/Library/FrameworkDoesNotSupportUIAutomationTests.cs (hunk 7 above)
Comment (DaveTryon) [reply src/RulesTest/Library/FrameworkDoesNotSupportUIAutomationTests.cs]: Removed (here and in the case that returns true)

**Review by RobGallo — APPROVED** [review]
